File: frontend/src/model/AssignmentValue.ts
import {SchoolLesson} from "./SchoolLesson";
import {Assignments, AssignmentType} from "./Assignments";
import {dateTimeToBulgarianLocaleString} from "../utils";
import {RoomToSubjects} from "./School";

export type AssignmentValue = ExaminationValue | HomeworkValue | EventValue


export class ExaminationValue {
    lesson: SchoolLesson

    exam: number | null

    constructor(lesson: SchoolLesson, exam: number | null) {
        this.lesson = lesson
        this.exam = exam
    }
}

export class HomeworkValue {
    to: string
    homeworkLesson: SchoolLesson

    constructor(to: string, homeworkLesson: SchoolLesson) {
        this.to = to
        this.homeworkLesson = homeworkLesson
    }
}

export class EventValue {
    from: string
    to: string
    room: RoomToSubjects

    constructor(from: string, to: string, room: RoomToSubjects) {
        this.from = from
        this.to = to
        this.room = room
    }
}

export const constructAssignmentValueMessage = (assignment: Assignments) => {
    let assignmentValueMessage = ''
    switch (assignment.assignmentType) {
        case AssignmentType.EXAMINATION: {
            assignmentValueMessage = ` - от ${(dateTimeToBulgarianLocaleString((assignment.assignmentValue as ExaminationValue).lesson.startTimeOfLesson))} до ${(dateTimeToBulgarianLocaleString((assignment.assignmentValue as ExaminationValue).lesson.endTimeOfLesson))}, място: ${(assignment.assignmentValue as ExaminationValue).lesson.room.room} стая, по: ${(assignment.assignmentValue as ExaminationValue).lesson.subject.name}`
            break;
        }
        case AssignmentType.HOMEWORK: {
            assignmentValueMessage = ` - краен срок: ${(dateTimeToBulgarianLocaleString((assignment.assignmentValue as HomeworkValue).to))}, по: ${(assignment.assignmentValue as HomeworkValue).homeworkLesson.subject.name}`
            break;
        }
        case AssignmentType.EVENT: {
            assignmentValueMessage = ` - от ${(dateTimeToBulgarianLocaleString((assignment.assignmentValue as EventValue).from))} до ${(dateTimeToBulgarianLocaleString((assignment.assignmentValue as EventValue).to))}, място: ${(assignment.assignmentValue as EventValue).room.room} стая`
            break;
        }
    }
    return assignmentValueMessage
}

File: backend/src/main/kotlin/com/nevexis/backend/schoolManagement/assignments/AssignmentValue.kt
@file:UseSerializers(
    LocalDateTimeSerializer::class
)

package com.nevexis.backend.schoolManagement.assignments

import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.nevexis.backend.schoolManagement.school.RoomToSubjects
import com.nevexis.backend.schoolManagement.school_lessons.SchoolLesson
import com.nevexis.backend.serializers.LocalDateTimeSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonContentPolymorphicSerializer
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import java.time.LocalDateTime

@Serializable(with = AssignmentValueSerializer::class)
@JsonDeserialize(using = AssignmentValueDeserializer::class)
sealed class AssignmentValue {
    @Serializable
    data class ExaminationValue(
        val lesson: SchoolLesson,
        val exam: Int? = null
    ) : AssignmentValue()

    @Serializable
    data class HomeworkValue(
        val to: LocalDateTime,
        val homeworkLesson: SchoolLesson
    ) : AssignmentValue()

    @Serializable
    data class EventValue(
        val from: LocalDateTime,
        val to: LocalDateTime,
        val room: RoomToSubjects
    ) : AssignmentValue()
}

object AssignmentValueSerializer :
    JsonContentPolymorphicSerializer<AssignmentValue>(AssignmentValue::class) {
    override fun selectDeserializer(element: JsonElement) = when {
        "lesson" in element.jsonObject -> AssignmentValue.ExaminationValue.serializer()
        "homeworkLesson" in element.jsonObject -> AssignmentValue.HomeworkValue.serializer()
        "room" in element.jsonObject -> AssignmentValue.EventValue.serializer()
        else -> error("There are no other student evaluations")
    }
}

class AssignmentValueDeserializer : JsonDeserializer<AssignmentValue>() {
    override fun deserialize(
        p: com.fasterxml.jackson.core.JsonParser,
        ctxt: DeserializationContext
    ): AssignmentValue {
        val rootNode = p.codec.readTree<JsonNode>(p) ?: throw IllegalArgumentException("Unable to parse JSON")
        val evaluationNode = rootNode.path("assignmentValue")

        if (evaluationNode.isMissingNode) {
            throw IllegalArgumentException("Assignment value is missing in the JSON")
        }

        val evaluationJson = evaluationNode.toString()

        return when {
            evaluationJson.contains("lesson") -> Json.decodeFromString<AssignmentValue.ExaminationValue>(
                evaluationJson
            )

            evaluationJson.contains("homeworkLesson") -> Json.decodeFromString<AssignmentValue.HomeworkValue>(
                evaluationJson
            )

            else -> Json.decodeFromString<AssignmentValue.EventValue>(evaluationJson)
        }
    }
}